The Linux QB failure is

```
11:42:41,907 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/event/socket.c: In function ‘socket_watcher_start’:
11:42:41,907 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/event/socket.c:102:11: error: conversion to ‘uint16_t’ from ‘int’ may alter its value [-Werror=conversion]
```

which doesn't make sense because both `sin6_port` and `sin_port` should be the same type -- `in_port_t`.  Everywhere I've check, that _is_ a typedef to `uint16_t`.  However, this is happening on Ubuntu 12.04, which is EOL, so I can't see what version of glibc it was using.

The FreeBSD QB failure is

```
    funcs.serverstart('127.0.0.1:')  -- assign random port
    assert(string.match(funcs.serverlist()[1], '127.0.0.1:%d+'))
```

on the assert:

```
19:17:33,940 INFO  - # test/functional/eval/server_spec.lua @ 61
19:17:33,940 INFO  - # Failure message: test/functional/eval/server_spec.lua:70: bad argument #1 to 'match' (string expected, got nil)
19:17:33,940 INFO  - # stack traceback:
19:17:33,940 INFO  - #     test/functional/eval/server_spec.lua:70: in function <test/functional/eval/server_spec.lua:61>
```

It'd be interesting to see what `serverlist()` is actually returning.

--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -6462,11 +6462,20 @@ serverlist()						*serverlist()*
 			nvim --cmd "echo serverlist()" --cmd "q"
 <
 serverstart([{address}])				*serverstart()*
-		Opens a named pipe or TCP socket at {address} for clients to
-		connect to and returns {address}. If no address is given, it
-		is equivalent to: >
+		Opens a TCP socket (IPv4/IPv6), Unix domain socket (Unix),
+		or named pipe (Windows) at {address} for clients to connect
+		to and returns {address}.
+
+		If {address} contains `:`, a TCP socket is used. Everything in
+		front of the last occurrence of `:` is the IP or hostname,
+		everything after it the port. If the port is empty or `0`,
+		a random port will be assigned.
+
+		If no address is given, it is equivalent to: >
 			:call serverstart(tempname())
+
 < 		|$NVIM_LISTEN_ADDRESS| is set to {address} if not already set.
+
 							*--servername*
 		The Vim command-line option `--servername` can be imitated: >
 			nvim --cmd "let g:server_addr = serverstart('foo')"

--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -45,6 +45,12 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
+    if (*port == NUL) {
+      // When no port is given, (uv_)getaddrinfo expects NULL otherwise the
+      // implementation may attempt to lookup the service by name (and fail)
+      port = NULL;
+    }
+
     uv_getaddrinfo_t request;
 
     int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,

--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -18,13 +18,14 @@
 #include "nvim/memory.h"
 #include "nvim/macros.h"
 #include "nvim/charset.h"
+#include "nvim/log.h"
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
 # include "event/socket.c.generated.h"
 #endif
 
-void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
-                         const char *endpoint)
+int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+                        const char *endpoint)
   FUNC_ATTR_NONNULL_ALL
 {
   xstrlcpy(watcher->addr, endpoint, sizeof(watcher->addr));
@@ -32,33 +33,34 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
   char *host_end = strrchr(addr, ':');
 
   if (host_end && addr != host_end) {
-    intmax_t port;
-    int ret = getdigits_safe(&(char_u *){ (char_u *)host_end + 1 }, &port);
-    if (ret == FAIL || port < 0 || port > UINT16_MAX) {
-      // Invalid port.
-      goto do_pipe;
+    // Split user specified address into two strings, addr(hostname) and port.
+    // The port part in watcher->addr will be updated later.
+    *host_end = '\0';
+    char *port = host_end + 1;
+    intmax_t iport;
+
+    int ret = getdigits_safe(&(char_u *){ (char_u *)port }, &iport);
+    if (ret == FAIL || iport < 0 || iport > UINT16_MAX) {
+      ELOG("Invalid port: %s", port);
+      return UV_EINVAL;
     }
 
-    *host_end = '\0';
     uv_getaddrinfo_t request;
 
-    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, host_end+1,
+    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
                                 &(struct addrinfo){
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
-    *host_end = ':';
     if (retval != 0) {
-      // Failed to look up address.
-      goto do_pipe;
+      ELOG("Host lookup failed: %s", endpoint);
+      return retval;
     }
-    *(host_end + 1) = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
 
     uv_tcp_init(&loop->uv, &watcher->uv.tcp.handle);
     watcher->stream = STRUCT_CAST(uv_stream_t, &watcher->uv.tcp.handle);
   } else {
-do_pipe:
     uv_pipe_init(&loop->uv, &watcher->uv.pipe.handle, 0);
     watcher->stream = STRUCT_CAST(uv_stream_t, &watcher->uv.pipe.handle);
   }
@@ -68,6 +70,8 @@ do_pipe:
   watcher->close_cb = NULL;
   watcher->events = NULL;
   watcher->data = NULL;
+
+  return 0;
 }
 
 int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
@@ -98,7 +102,7 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
           : ((struct sockaddr_in6 *)&sas)->sin6_port;
         // v:servername uses the string from watcher->addr
         size_t len = strlen(watcher->addr);
-        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, "%" PRIu16,
+        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, ":%" PRIu16,
                  ntohs(port));
         break;
       }

--- a/src/nvim/msgpack_rpc/server.c
+++ b/src/nvim/msgpack_rpc/server.c
@@ -118,7 +118,12 @@ int server_start(const char *endpoint)
   }
 
   SocketWatcher *watcher = xmalloc(sizeof(SocketWatcher));
-  socket_watcher_init(&main_loop, watcher, endpoint);
+
+  int result = socket_watcher_init(&main_loop, watcher, endpoint);
+  if (result < 0) {
+    xfree(watcher);
+    return result;
+  }
 
   // Check if a watcher for the endpoint already exists
   for (int i = 0; i < watchers.ga_len; i++) {
@@ -132,7 +137,7 @@ int server_start(const char *endpoint)
     }
   }
 
-  int result = socket_watcher_start(watcher, MAX_CONNECTIONS, connection_cb);
+  result = socket_watcher_start(watcher, MAX_CONNECTIONS, connection_cb);
   if (result < 0) {
     ELOG("Failed to start server: %s", uv_strerror(result));
     socket_watcher_close(watcher, free_server);

